### PR TITLE
Rework the wait for healthy logic to count down nodes for CASSANDRA-14297

### PR DESCRIPTION
**Pull request by jolynch**
This improves on the wait for healthy work from CASSANDRA-13993 to
solve CASSANDRA-14297. It allows the cluster owners to fine tune the
wait for behaviour to ensure availability of their application during
rolling restarts. The defaults are to wait for all but one local DC host
to be alive and up, and not care about remote DC hosts.

### CODE REVIEW

**Review by sumanth-pasupuleti — APPROVED** [review]

File: src/java/org/apache/cassandra/config/Config.java [hunk 1]
```diff
@@ -376,9 +376,31 @@
...
+     *     LOCAL_QUORUM      | 1            | default (any)
+     *     ONE               | any          | RF - 1
+     *     QUORUM            | any          | (RF / 2) - 1
+     *     ALL               | default      | 0
+     *
+     * A concrete example with QUORUM would be if you have 3 replicas in 2 datacenters, then you would set
+     * block_for_peers_all_dcs to (6 / 2) - 1 = 2 because that guarantees that at most 2 hosts in all datacenters
+     * are down when you start taking client traffic, which should satistfy QUORUM for all RF=6 QUORUM queries.
+     */
+    public int block_for_peers_local_dc = 2;
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/config/Config.java]: If the goal is to not return unavailable shouldn't the default really be 1 since LOCAL_QUORUM and QUORUM are more common? I feel like if this doesn't do what people want with the defaults it makes it a lot less useful?

All this does is add 10 seconds to startup which seems really minor. A part of me even wonders why fuss with any of this if we are just going to wait 10 seconds? Just wait until everything is up and if it doesn't happen in 10 seconds continue to do what we were going to do anyways?

The connection priming stuff I get. That is good and I'm glad we added that.

The name also doesn't really make sense anymore since this is not the number we are blocking for it's the number we aren't blocking for.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java [hunk 2]
```diff
@@ -39,7 +42,9 @@
 import org.apache.cassandra.locator.InetAddressAndPort;
 import org.apache.cassandra.net.async.OutboundConnectionIdentifier.ConnectionType;
 import org.apache.cassandra.utils.FBUtilities;
+import org.jboss.byteman.synchronization.CountDown;
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Unused import

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java [hunk 3]
```diff
@@ -73,56 +80,84 @@ public static StartupClusterConnectivityChecker create(int targetPercent, int ti
      * @return true if the requested percentage of peers are marked ALIVE in gossip and have their connections opened;
      * else false.
      */
-    public boolean execute(Set<InetAddressAndPort> peers)
+    public boolean execute(Set<InetAddressAndPort> peers, Function<InetAddressAndPort, String> getDatacenter)
     {
-        if (targetPercent == 0 || peers == null)
+        if (peers == null)
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: This change appears to remove the ability to disable this? It's an interesting change because it means we will always prime the connections which sounds kind of useful.

The most you can do is crank down the timeout or set the number of hosts you won't wait for very high which has a similar impact.

File: src/java/org/apache/cassandra/config/Config.java [hunk 4]
```diff
@@ -376,9 +376,31 @@
...
+    public int block_for_peers_all_dcs = Integer.MAX_VALUE;
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/config/Config.java]: QUORUM is at least a little common? Seems like defaulting to no waiting for remote DCs might not be the right default? It's a question of which do people dislike more? Getting unavailables when a node restarts or potentially waiting extra time on startup?

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java [hunk 5]
```diff
@@ -73,56 +80,84 @@ public static StartupClusterConnectivityChecker create(int targetPercent, int ti
...
 
-        logger.info("choosing to block until {}% of the {} known peers are marked alive and connections are established; max time to wait = {} seconds",
-                    targetPercent, peers.size(), TimeUnit.NANOSECONDS.toSeconds(timeoutNanos));
+        Set<InetAddressAndPort> myLocalPeers = peers.stream()
+                                                    .collect(Collectors.groupingBy(getDatacenter, toSet()))
+                                                    .getOrDefault(getDatacenter.apply(FBUtilities.getBroadcastAddressAndPort()),
+                                                                  Collections.emptySet());
+
+
+        logger.info("choosing to block until no more than {}/{} local and no more than {}/{} global peers are still DOWN; max time to wait = {} seconds",
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: This logs the local node as a peer since you aren't removing it. Seems like it's not a desired side effect.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java [hunk 6]
```diff
@@ -17,13 +17,16 @@
...
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.stream.Collectors;
 
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.Sets;
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Unused import

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java [hunk 7]
```diff
@@ -213,4 +261,11 @@ boolean incrementAndCheck(InetAddressAndPort address)
             return acks.computeIfAbsent(address, addr -> new AtomicInteger(0)).incrementAndGet() == threshold;
         }
     }
+
+    private String fmtBlocker(Integer size, Integer count)
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Is this a more complicated Math.max?

**Review by aweisberg — CHANGES_REQUESTED** [review]
Some minor nits. 

At this point I question why we don't just wait until all nodes are up since the maximum amount of time we will wait is 10 seconds. It seems like going faster than that in the case where one or two nodes are down doesn't sound so useful?

I wonder if things could be simpler that way.

File: src/java/org/apache/cassandra/config/Config.java (hunk 1 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/config/Config.java]: > If the goal is to not return unavailable shouldn't the default really be 1 since LOCAL_QUORUM and QUORUM are more common? I feel like if this doesn't do what people want with the defaults it makes it a lot less useful?

I thought that defaulting to handling the case that the drivers default to `LOCAL_ONE` was most sensible, although I can make it `1` if you prefer and say the defaults are for `LOCAL_QUORUM` (I don't have strong opinions other than we shouldn't be waiting by default on remote DCs).

> All this does is add 10 seconds to startup which seems really minor. A part of me even wonders why fuss with any of this if we are just going to wait 10 seconds? Just wait until everything is up and if it doesn't happen in 10 seconds continue to do what we were going to do anyways?

In practice you wait way less than 10 seconds (e.g. on the test 200 node 4.0 cluster we handshake with the whole local DC in about ... 500ms). I personally think that most users would rather their database wait O(minutes) than throw errors in the general case, but from the earlier conversations in CASSANDRA-13993 it seemed like folks were hesitant to wait so long on startup e.g. when doing ccm clusters or the such.

> The name also doesn't really make sense anymore since this is not the number we are blocking for it's the number we aren't blocking for.

How about ... `startup_max_down_local_dc_peers` and `startup_max_down_peers`?

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 2 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: ack

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 3 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Yea that was my idea, that we should always prime the connections but we can just not wait (if you set the local dc option to a really large number then it basically primes and immediately returns).

File: src/java/org/apache/cassandra/config/Config.java (hunk 4 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/config/Config.java]: I agree and 10s isn't that big a deal (imo) for users that don't want to wait for remote dcs, but I was trying to compromise between the old settings (70% == could literally have a whole DC down with 3 DCs) which appear to be motivated from a perspective of "don't block my startup" and what I personally think that we shouldn't be saying we're ready to coordinate until we're actually ready to coordinate.

I am happy to put whatever default gets this merged ;-) We'll probably internally be setting the local setting to `1` (so `LOCAL_QUORUM`) and the remote to a really large number. But that's just our perspective...

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 5 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Yea ... I was debating if this should be here or down below 122 where we remove it. I was thinking that users would want to see their cluster size in the denominator (e.g. `choosing to block until no more than 1/200` nodes) even if strictly speaking we just immediately succeed on the local host.

I don't have strong preferences, I'll take the local node out of the count.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 6 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: ack

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 7 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Heh, yea it is I just thought that it was slightly clearer what it was doing (so we don't print `Integer.MAX_VALUE` to the log. I'll change it to `Math.max` instead.

File: src/java/org/apache/cassandra/config/Config.java (hunk 4 above)
Comment (aweisberg) [reply src/java/org/apache/cassandra/config/Config.java]: This one of those things where I want to ask around and see if we can get away with just waiting 10 seconds if a node is down.

Seems like the solution for CCM is to just have CCM set the config it wants. I'm also wondering if CCM actually waits the 10 seconds anyways just because it starts one node at a time and waits for a specific log message a lot of the time and I'm not sure if that message is before or after the connectivity checker.

I have to go take a look at how CCM works.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 3 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: In the latest version I added the ability to disable this by setting a negative timeout. I guess it's probably a good idea to have an opt out option since we're post freeze.

File: src/java/org/apache/cassandra/config/Config.java (hunk 4 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/config/Config.java]: I played around with CCM and it appears to work fine (starting a cluster, stopping it, starting and stopping nodes etc). I think we should be good on this front.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 5 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: I changed it to remove the local peer like before :-)

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 7 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: I killed it.

**Comment by jolynch** [comment]
Ok, I've rebased and incorporated the feedback from IRC/jira as well. Please let me know if there are further changes I need to make.

File: src/java/org/apache/cassandra/config/Config.java [hunk 8]
```diff
@@ -380,9 +380,23 @@
     public RepairCommandPoolFullStrategy repair_command_pool_full_strategy = RepairCommandPoolFullStrategy.queue;
     public int repair_command_pool_size = concurrent_validations;
 
-    // parameters to adjust how much to delay startup until a certain amount of the cluster is connect to and marked alive
-    public int block_for_peers_percentage = 70;
+    /**
+     * When a node first starts up it intially considers all other peers as DOWN, and then as the initial gossip
+     * broadcast messages comes back nodes transition to UP. These options configure how long we wait for peers to
+     * connect before we make this node available as a coordinator. Furthermore, if this feature is enabled
+     * (timeout >= 0) Cassandra initiates the non gossip channel internode connections on startup as well and waits
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/config/Config.java]: This part of the description doesn't match the description for block_for_peers_timeot_in_secs. It sounds like now it differentiates between disabled, and prime the connections.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java [hunk 9]
```diff
@@ -48,81 +51,133 @@
...
-    private final int targetPercent;
+    private final boolean blockForRemoteDcs;
     private final long timeoutNanos;
 
-    public static StartupClusterConnectivityChecker create(int targetPercent, int timeoutSecs)
+    public static StartupClusterConnectivityChecker create(long timeoutSecs, boolean blockForRemoteDcs)
     {
-        timeoutSecs = Math.max(1, timeoutSecs);
+        if (timeoutSecs < 0)
+            logger.warn("skipping block-for-peers due to negative timeout. You may encounter errors or timeouts on" +
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Should this be a warning? They have pretty clearly asked for it?

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java [hunk 10]
```diff
@@ -48,81 +51,133 @@
...
+        String localDc = getDatacenterSource.apply(localAddress);
 
+        peers.remove(localAddress);
         if (peers.isEmpty())
             return true;
 
-        logger.info("choosing to block until {}% of the {} known peers are marked alive and connections are established; max time to wait = {} seconds",
-                    targetPercent, peers.size(), TimeUnit.NANOSECONDS.toSeconds(timeoutNanos));
+        // make a copy of the datacenter mapping (in case gossip updates happen during this method or some such)
+        Map<InetAddressAndPort, String> datacenterMap = peers.stream()
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Bikeshedding, but datacenterMap is not very descriptive. What it is is a peerToDatacenter. You can also omit mpa it's kind of redundant with XtoY which automatically implies it's a map relationship. Multimaps are handles by making Y plural. I really think it's worth using better naming here.

Also pulling out getDataCenter is a few characters more succinct, but I would rather see XtoY::get repeated a few times. If it's not referring to the source then the name also needs to be more representative like getDatacenterFromPeer which is basically just another way of saying peerToDatacenter::get.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java [hunk 11]
```diff
@@ -48,81 +51,133 @@
...
+                                                             .collect(Collectors.toMap(k -> k, getDatacenterSource));
+        Function<InetAddressAndPort, String> getDatacenter = datacenterMap::get;
 
-        long startNanos = System.nanoTime();
+        Map<String, Set<InetAddressAndPort>> peersByDc = peers.stream()
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Could this be a set multimap?

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java [hunk 12]
```diff
@@ -48,81 +51,133 @@
...
             if (Gossiper.instance.isAlive(peer) && alivePeers.add(peer) && acks.incrementAndCheck(peer))
-                latch.countDown();
+            {
+                String datacenter = getDatacenter.apply(peer);
+                if (latchMap.containsKey(datacenter))
+                    latchMap.get(datacenter).countDown();
+            }
+        }
+
+        boolean succeeded = Uninterruptibles.awaitUninterruptibly(latchMap.get(localDc), timeoutNanos, TimeUnit.NANOSECONDS);
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Why do you need the loop and this? Since you removed other DCs from peersByDC shouldn't there already be only the local DC in the latch map?

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java [hunk 13]
```diff
@@ -48,81 +51,133 @@
...
+        for (String datacenter: latchMap.keySet())
+        {
+            if (datacenter.equals(localDc))
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Then you don't need this exception?

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java [hunk 14]
```diff
@@ -48,81 +51,133 @@
...
+        else
+        {
+            logger.info("Blocking coordination until only a single peer is DOWN in each datacenter, timeout={}s",
+                        TimeUnit.NANOSECONDS.toSeconds(timeoutNanos));
+        }
 
         AckMap acks = new AckMap(3);
-        int target = (int) ((targetPercent / 100.0) * peers.size());
-        CountDownLatch latch = new CountDownLatch(target);
+        Map<String, CountDownLatch> latchMap = new HashMap<>(peersByDc.size());
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Another candidate for datacenterToY. Not quite sure of the perfect name for Y. I am fine with using dc if you want shorter.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java [hunk 15]
```diff
@@ -48,81 +51,133 @@
...
+        Function<InetAddressAndPort, String> getDatacenter = datacenterMap::get;
 
-        long startNanos = System.nanoTime();
+        Map<String, Set<InetAddressAndPort>> peersByDc = peers.stream()
+                                                              .collect(Collectors.groupingBy(getDatacenter,
+                                                                                             Collectors.toSet()));
+
+        if (!blockForRemoteDcs)
+        {
+            peersByDc.keySet().retainAll(Collections.singleton(localDc));
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: This important bit of behavior is tucked away here and I only know about it because I went looking. Can you add a mostly redundant comment mentioning that you remove all but the local DC so that we only wait on nodes in the local DC?

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java [hunk 16]
```diff
@@ -48,81 +51,133 @@
...
         // set up a listener to react to new nodes becoming alive (in gossip), and account for all the nodes that are already alive
-        Set<InetAddressAndPort> alivePeers = Sets.newSetFromMap(new ConcurrentHashMap<>());
-        AliveListener listener = new AliveListener(alivePeers, latch, acks);
+        Set<InetAddressAndPort> alivePeers = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        AliveListener listener = new AliveListener(alivePeers, latchMap, acks, getDatacenter);
         Gossiper.instance.register(listener);
 
         // send out a ping message to open up the non-gossip connections
-        sendPingMessages(peers, latch, acks);
+        sendPingMessages(peers, latchMap, acks, getDatacenter);
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Comment that pings are sent to all peers even if latchMap doesn't contain a latch for them.

File: test/unit/org/apache/cassandra/net/StartupClusterConnectivityCheckerTest.java [hunk 17]
```diff
@@ -36,13 +36,34 @@
...
+
+    private String getDatacenter(InetAddressAndPort endpoint)
+    {
+        if (peersA.contains(endpoint))
+            return "datacenterA";
+        if (peersB.contains(endpoint))
+            return "datacenterB";
+        else if (peersC.contains(endpoint))
+            return "datacenterC";
+        return "NA";
```
Comment (aweisberg) [inline test/unit/org/apache/cassandra/net/StartupClusterConnectivityCheckerTest.java]: Should probably return null

File: test/unit/org/apache/cassandra/net/StartupClusterConnectivityCheckerTest.java [hunk 18]
```diff
@@ -69,33 +113,102 @@ public void tearDown()
...
+        // local peer plus 4 peers from same dc should pass (5/6)
+        available.clear();
+        copyCount(peersAMinusLocal, available, NUM_PER_DC - 2);
+        checkAvailable(localQuorumConnectivityChecker, available, true, true);
+    }
+
+    @Test
+    public void execute_GlobalQuorum()
+    {
+        // local dc passing shouldn'nt pass globally with two hosts down in datacenterB
```
Comment (aweisberg) [inline test/unit/org/apache/cassandra/net/StartupClusterConnectivityCheckerTest.java]: typo

File: test/unit/org/apache/cassandra/net/StartupClusterConnectivityCheckerTest.java [hunk 19]
```diff
@@ -69,33 +113,102 @@ public void tearDown()
...
+    @Test
+    public void execute_Noop()
+    {
+        checkAvailable(noopChecker, new HashSet<>(), true, false);
+    }
+
+    @Test
+    public void execute_ZeroWait()
+    {
+        checkAvailable(zeroWaitChecker, new HashSet<>(), false, false);
```
Comment (aweisberg) [inline test/unit/org/apache/cassandra/net/StartupClusterConnectivityCheckerTest.java]: Maybe add a test that does zero wait, but also checks the connections are made?

**Review by aweisberg — CHANGES_REQUESTED** [review]
Very close! Mostly bikeshedding.

File: src/java/org/apache/cassandra/config/Config.java (hunk 8 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/config/Config.java]: Ok, I reworded it. I think that having an escape hatch makes sense, what do you think?

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 9 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: I thought it should be for the same reason we warn them if they set the timeout to more than 100, but I don't care too much. I removed it.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 10 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Sure thing, makes sense and I've made both refactors.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 11 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Sure, done.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 12 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: I see what you're saying about just iterating in any order, sg2m.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 13 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Makes sense, removed

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 14 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: I called it `dcToRemainingPeers` to try to convey what it was counting down. Let me know if that's not clear and I'll change it.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 15 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Ack, added the comment.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 16 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: Done.

File: test/unit/org/apache/cassandra/net/StartupClusterConnectivityCheckerTest.java (hunk 17 above)
Comment (jolynch) [reply test/unit/org/apache/cassandra/net/StartupClusterConnectivityCheckerTest.java]: Right, so it'll fail fast, done.

File: test/unit/org/apache/cassandra/net/StartupClusterConnectivityCheckerTest.java (hunk 18 above)
Comment (jolynch) [reply test/unit/org/apache/cassandra/net/StartupClusterConnectivityCheckerTest.java]: Ack, fixed.

File: test/unit/org/apache/cassandra/net/StartupClusterConnectivityCheckerTest.java (hunk 19 above)
Comment (jolynch) [reply test/unit/org/apache/cassandra/net/StartupClusterConnectivityCheckerTest.java]: Makes sense, I refactored this test to look at that instead of just returning immediately.

File: src/java/org/apache/cassandra/config/Config.java (hunk 1 above)
Comment (jolynch) [reply src/java/org/apache/cassandra/config/Config.java]: We ended up going the direction of a boolean.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java (hunk 9 above)
Comment (aweisberg) [reply src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java]: The reason I thought it was strange was because -1 was a valid config option.

File: src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java [hunk 20]
```diff
@@ -48,81 +53,135 @@
...
+        sendPingMessages(peers, dcToRemainingPeers, acks, peerToDatacenter::get);
 
         for (InetAddressAndPort peer : peers)
+        {
             if (Gossiper.instance.isAlive(peer) && alivePeers.add(peer) && acks.incrementAndCheck(peer))
-                latch.countDown();
+            {
+                String datacenter = peerToDatacenter.get(peer);
+                // We have to check because we might only have the local DC in the map
+                if (dcToRemainingPeers.containsKey(datacenter))
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java:146]: I know some people who are totally going to be triggered by this since it does the lookup twice, but it is 100% fine. I think it's clearer this way. Get with null check is three lines instead of two. Really I just mention it so later if they come back and complain I can claim it was intentional :-)

**Comment by jolynch** [comment]
Closing as this was committed in https://github.com/apache/cassandra/commit/801cb70ee811c956e987718a00695638d5bec1b6
